### PR TITLE
Implemented JDBC

### PR DESCRIPTION
**Pull request by KovalAlexey**

### CODE REVIEW

File: src/main/resources/init_db.sql [hunk 1]
```diff
@@ -0,0 +1,8 @@
+CREATE SCHEMA `online_store` DEFAULT CHARACTER SET utf8 ;
+
+CREATE TABLE `internet-shop`.`products` (
+  `product_id` BIGINT(11) NOT NULL AUTO_INCREMENT,
+  `productName` VARCHAR(225) NOT NULL,
+  `price` DOUBLE NOT NULL,
+  `deleted` TINYINT NULL,
+  PRIMARY KEY (`product_id`));
```
Comment (margarita-ignatieva) [inline src/main/resources/init_db.sql]: empty line

**Comment by margarita-ignatieva** [comment]
Do you check for ProductDaoJdbcImpl work somewhere in the test class (where you used ServiceImpl earlier)?

**Review by margarita-ignatieva — APPROVED** [review]

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.List;
+import java.util.Optional;
+
+@Dao
+public class ProductDaoJdbcImpl implements ProductDao {
+    @Override
+    public Product create(Product product) {
+        String query = "INSERT INTO products (productName, price) VALUES (?, ?)";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            statement.setString(1, product.getName());
+            statement.setDouble(2, product.getPrice());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                product.setId(resultSet.getLong(1));
+            }
+            return product;
+        } catch (SQLException e) {
+            throw new DataProcessException("Can't create product!", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: let's add product name to messsage to make more informative

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        }
+    }
+
+    @Override
+    public Optional<Product> get(Long productId) {
+        String query = "SELECT * FROM products WHERE product_id = ? AND deleted != 1";
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
        String query = "SELECT * FROM products WHERE product_id = ? AND deleted = FALSE";
```
this way query would be more readable 
same with other queries

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, productId);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                Product product = productFromSet(resultSet);
+                product.setId(productId);
+                return Optional.of(product);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessException("Can't find product with id", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: add id to message

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    @Override
+    public List<Product> getAll() {
+        String query = "SELECT * FROM products WHERE deleted != 1";
+        List<Product> allProducts = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            ResultSet resultSet = statement.executeQuery(query);
+            while (resultSet.next()) {
+                Product product = productFromSet(resultSet);
+                product.setId(resultSet.getLong("product_id"));
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: can we move this line to our private method?

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, productId);
+            statement.executeUpdate();
+            return true;
+        } catch (SQLException e) {
+            throw new DataProcessException("Can't execute update query!", e);
+        }
+    }
+
+    private Product productFromSet(ResultSet resultSet) {
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
    private Product getProductFromSet(ResultSet resultSet) {
```
don't forget about method naming convention

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    public Product update(Product product) {
+        String query = "UPDATE products SET productName = ?, price = ? WHERE product_id = ?";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setString(1, product.getName());
+            statement.setDouble(2, product.getPrice());
+            statement.setLong(3, product.getId());
+            statement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessException("Can't update product", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: add id to message

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        }
+        return product;
+    }
+
+    @Override
+    public boolean delete(Long productId) {
+        String query = "UPDATE products SET deleted = true WHERE product_id = ?";
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
        String query = "UPDATE products SET deleted = TRUE WHERE product_id = ?";
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, productId);
+            statement.executeUpdate();
+            return true;
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: it's not very good idea to return true always 
let's return statement.executeUpdate() == 1 for example

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        } catch (SQLException e) {
+            throw new DataProcessException("Can't execute update query!", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessException("Can't execute update product!", e);
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        }
+    }
+
+    private Product productFromSet(ResultSet resultSet) {
+        try {
+            String productName = resultSet.getString("productName");
+            double price = resultSet.getDouble("price");
+            return new Product(productName, price);
+        } catch (SQLException e) {
+            throw new DataProcessException("Can't generate product...", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessException("Can't retrieve product from ResultSet ", e);
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 7 above)
Comment (KovalAlexey) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 9 above)
Comment (KovalAlexey) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 4 above)
Comment (KovalAlexey) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 2 above)
Comment (KovalAlexey) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 3 above)
Comment (KovalAlexey) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 5 above)
Comment (KovalAlexey) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 8 above)
Comment (KovalAlexey) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 11 above)
Comment (KovalAlexey) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 12 above)
Comment (KovalAlexey) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/resources/init_db.sql (hunk 1 above)
Comment (KovalAlexey) [reply src/main/resources/init_db.sql]: done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 10 above)
Comment (KovalAlexey) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 6 above)
Comment (KovalAlexey) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 8 above)
Comment (RomanDubovskyi) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: this one is not done yet

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+            statement.setLong(1, productId);
+            return statement.executeUpdate() == 1;
+        } catch (SQLException e) {
+            throw new DataProcessException("Can't execute update product " + productId, e);
+        }
+    }
+
+    private Product getProductFromSet(ResultSet resultSet) {
+        try {
+            String productName = resultSet.getString("productName");
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            String productName = resultSet.getString("product_name");
```
use same naming style

Comment (gradiviro) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java:103]: if name should differ i would change String var name, cuz  `productName` its my column name in DB

Comment (KovalAlexey) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+        List<Product> allProducts = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            ResultSet resultSet = statement.executeQuery(query);
+            while (resultSet.next()) {
+                Product product = getProductFromSet(resultSet);
+                allProducts.add(product);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessException("Can't execute query!", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessException("Can't get all products!", e);
```

**Review by RomanDubovskyi — APPROVED** [review]

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 14 above)
Comment (KovalAlexey) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: done

**Pull request merged**